### DISCUSSION TITLE
Variable window length

### DISCUSSION THREAD

**Discussion by krisvanneste** [SeismicSource/sourcespec#46] [Ideas]
Hi Claudio,

I have started looking how a variable window length could be implemented in sourcespec.
One idea is simply to define the window length variability as a fraction of the travel time of the considered wave.
This could be implemented easily as follows:

1. add a configuration parameter: `variable_win_length_factor` (or maybe `variable_win_length_tt_fraction` to be more explicit) with default value 0.
2. add following code in ssp_process_traces._define_signal_and_noise_windows:
```
- tt_[p/s] = v[p/s]_arrival_time - trace.stats.event.hypocenter.origin_time
- win_length_[p/s] = max(config.win_length, variable_win_length_factor*tt_[p/s])
```
If origin time is not set in the trace header (or maybe it is guaranteed to be set when we get to this function?), we can calculate travel time in a different way:
```
- v[p/s]_tt = config.v[p/s]_tt or (v[p/s]_source + v[p/s]_stations) / 2.
- tt_[p/s] = trace.stats.hypo_dist / v[p/s]_tt
```
Either way, all required information is available to the _define_signal_and_noise_windows function.
I think 0.25 to 0.5 would be a reasonable range for the variable_window_length_factor, but we can probably test this.
We may also need to check if the window length does not extend beyond the trace length (but maybe this is not a problem).

Further, to determine the noise window length, we need to select `win_length_[p/s]` depending on `config.wave_type`
We should probably also warn users that variable win_length also requires longer noise windows, particularly if noise weighting is used. This may imply adjusting `config.noise_pre_time` ...

Let me know what you think,
Kris

**Comment by krisvanneste** [comment]
@claudiodsf Do you have time to look at this?

**Comment by claudiodsf** [comment]
Hi Kris, thanks for the reminder and for this proposition!

Let me try to make some quick comments right now 😉 :

- I agree with the general idea 👍 
- travel time is always defined; otherwise an error is generated [here](https://github.com/SeismicSource/sourcespec/blob/878e9681ac47c838e93dd0bdc5eb57179d1b98da/sourcespec/ssp_process_traces.py#L268)
- a travel time dict is stored into trace objects: `trace.stats.travel_times`. The dict keys are the phase names. See [here](https://github.com/SeismicSource/sourcespec/blob/878e9681ac47c838e93dd0bdc5eb57179d1b98da/sourcespec/ssp_wave_arrival.py#L227)
- origin time can be undefined, see [here](https://github.com/SeismicSource/sourcespec/blob/878e9681ac47c838e93dd0bdc5eb57179d1b98da/sourcespec/ssp_read_sac_header.py#L111)
- a warning is issued if the noise window ends after the P arrival, see [here](https://github.com/SeismicSource/sourcespec/blob/878e9681ac47c838e93dd0bdc5eb57179d1b98da/sourcespec/ssp_process_traces.py#L307)
- I would prefer to zero-pad the variable time windows (signal and noise) to the same length for all the stations (maybe, making the parameter `spectral_win_length` mandatory when `variable_win_length_tt_fraction` is used?). This will make it possible to compare spectra from different earthquakes at the same station, especially useful when staking residuals. 

Let me know if that is enough to start drafting an implementation and if you can do it 😉 .

Claudio

**Reply by krisvanneste** [reply]
Claudio,

Thank you. I will give it a try on Thursday.
Kris
